numba/core/typeinfer.py

```diff
@@ -1479,6 +1479,17 @@ def typeof_global(self, inst, target, gvar):
             # Global array in nopython mode is constant
             typ = typ.copy(readonly=True)
 
+        if isinstance(typ, types.Tuple):
+            if all(isinstance(ty, (types.Integer, types.UnicodeType))
+                   for ty in typ):
+                typ = types.Tuple([types.literal(val) for val in gvar.value])
+
+        if isinstance(typ, types.UniTuple):
+            # since this test makes typ into a Tuple, it must be placed after
+            # the test for types.Tuple to prevent double-triggering
+            if isinstance(typ.dtype, (types.Integer, types.UnicodeType)):
+                typ = types.Tuple([types.literal(val) for val in gvar.value])
```

Comment:
I think your question is whether the two `if` statements could be merged, right? They could, but in exchange for a small inefficiency: not checking `dtype` of `UniTuple` and running instead the `all(...)` test on both types, like this:
```python 
        if isinstance(typ, types.BaseAnonymousTuple):
            types_with_literals = (types.Integer, types.UnicodeType)
            if all(isinstance(ty, types_with_literals) for ty in typ):
                typ = types.Tuple([types.literal(val) for val in gvar.value])
```
If we are not expecting tuples with 1000s of items, then this should not be a problem. 
Alternatively, if you'd like to have only one `if` for both tuples, and be as efficient as possible, we could have something like
```python
        if isinstance(typ, types.BaseAnonymousTuple):
            types_w_literals = (types.Integer, types.UnicodeType)
            types_in_tuple = (typ.dtype.types 
                              if isinstance(typ.dtype, types.UnionType) 
                              else (typ.dtype,))
            if all(isinstance(ty, types_w_literals) for ty in types_in_tuple):
                typ = types.Tuple([types.literal(val) for val in gvar.value])
```
That one is based on the `dtype` of `Tuple` being a `Union` from which all the unique types can be found in `dtype.types` without having to explicitly loop over them.
If small performance gains are important, I would go with the last version, at the cost of decreased readibility. If not, the first one is better.

Reply:
Ah, apologies, this wasn't clear, so what I should have said is as follows: 

* In the proposed code there's a required ordering in test conditions, first for Tuple, then for UniTuple. This is "ok", but there's more likelihood it gets accidentally inverted or refactored etc as it's just bare `if` statements in the block, not `elif`s.
* If the above was written so that there was a branch on e.g. `_BaseTuple` at the block level and then ordering was captured within that branch, it might prevent future accidents. IIRC there's a limit to tuple size anyway as it makes the stack enormous.

It's not a big problem either way, what do you think?

Reply:
ok, I have a small preference for avoid long loops so I'll go with the second option.
I couldn't find any reference to `_BaseTuple` in the code. Do you mean `BaseAnonymousTuple` or did I miss something?

Reply:
Thanks for adding the fix. Yes, `BaseAnonymousTuple` is fine, I was remembering incorrectly.